statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/backpressure/ThresholdBackPressureValve.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.statefun.flink.core.backpressure;
+
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashMap;
+import java.util.Objects;
+import org.apache.flink.statefun.sdk.Address;
+
+/** A simple Threshold based {@link BackPressureValve}. */
+public final class ThresholdBackPressureValve implements BackPressureValve {
+  private final int maximumPendingAsynchronousOperations;
+
+  /**
+   * a set of address that had explicitly requested to stop processing any new inputs (via {@link
+   * AsyncWaiter#awaitAsyncOperationComplete()}. Note that this is a set implemented on top of a
+   * map, and the value (Boolean) has no meaning.
+   */
+  private final ObjectOpenHashMap<Address, Boolean> blockedAddressSet =
+      new ObjectOpenHashMap<>(1024);
+
+  private int pendingAsynchronousOperationsCount;
+
+  /**
+   * Constructs a ThresholdBackPressureValve.
+   *
+   * @param maximumPendingAsynchronousOperations the total allowed async operations to be inflight
+   *     per StreamTask, or {@code -1} to disable back pressure.
+   */
+  public ThresholdBackPressureValve(int maximumPendingAsynchronousOperations) {
+    this.maximumPendingAsynchronousOperations = maximumPendingAsynchronousOperations;
+  }
+
+  public boolean shouldBackPressure() {
+    return totalPendingAsyncOperationsAtCapacity() || hasBlockedAddress();
+  }
+
+  public void blockAddress(Address address) {
+    Objects.requireNonNull(address);
+    blockedAddressSet.put(address, Boolean.TRUE);
+  }
+
+  public void notifyAsyncOperationRegistered() {
+    pendingAsynchronousOperationsCount++;
+  }
+
+  public void notifyAsyncOperationCompleted(Address owningAddress) {
+    Objects.requireNonNull(owningAddress);
+    pendingAsynchronousOperationsCount--;
+    blockedAddressSet.remove(owningAddress);
```

Comment:
I'm slightly confused here, could you help me a bit with clarifying?:

is `notifyAsyncOperationCompleted` meant to be called when ALL async operations from `owningAddress` has completed, or when SOME async operation from `owningAddress` has completed?

From  the `blockedAddressSet.remove(owningAddress)` operation, it seems like its supposed to be the former, but the `pendingAsynchronousOperationsCount` counter keeps track of in-flight asyncs across all addresses, and is not being checked before the address removal.

Reply:
Let me try to clarify:

There are two cases where a backpressure would be triggered:
a) The total number of in-flight async operations in a `StreamTask` exceeds a predefined threshold. This is tracked by `pendingAsynchronousOperationsCount`, it is incremented when an asyncOp is registered, and decremented when it is completed.
b) A specific address has requested to stop processing new inputs, this is tracked by the `blockedAddressSet`.

The method `notifyAsyncOperationCompleted(owningAddress)` is meant to be called when ANY async operation has been completed (with the address supplied as an argument). And the implementation needs to decide what to do with that information.
In the `ThresholdBackPressureValve` we need to account for (a) by decrement the total inflight async operations, and in addition, for (b) we check if that operation can unblock a previously blocked address.

I hope it helps, if it does let me reflect that comment in a JavaDoc of that class.

Reply:
@igalshilman
Thanks, that was very helpful.

Could you add to the Javadoc of `blockAddress` that it is unblocked as soon as some (one) async operation for the address completes?